### PR TITLE
Compile ulf

### PR DESCRIPTION
**Pull request by elafnm**
Greetings dears @playpaolo and @pooyan-dadvand ,

kindly please, find this pull request for the ULF application "with pybind",

thank you for your support and appreciated effort, 

best regards,

### CODE REVIEW

File: applications/FluidDynamicsApplication/python_scripts/navier_stokes_solver_vmsmonolithic.py
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/python_scripts/navier_stokes_solver_vmsmonolithic.py]: Why are you changing this solver?

File: .gitignore
Comment (rubenzorrilla) [inline .gitignore]: You should edit your local ignore file as explained in https://github.com/KratosMultiphysics/Kratos/wiki/How-to-make-Git-ignore-the-files-resulting-from-a-compilation-without-conflicts-in-.gitignore

File: .gitignore
Comment (rubenzorrilla) [inline .gitignore]: Indeed, I'm wondering why this is already here...

File: .gitignore
Comment (rubenzorrilla) [inline .gitignore]: Same...

File: applications/MeshingApplication/custom_python/add_meshers_to_python.cpp
Comment (rubenzorrilla) [inline applications/MeshingApplication/custom_python/add_meshers_to_python.cpp]: Follow the naming convention ReGenerateMeshDroplet.

File: applications/MeshingApplication/test_examples/dam3d_remesh.gid/out.out
Comment (rubenzorrilla) [inline applications/MeshingApplication/test_examples/dam3d_remesh.gid/out.out:2]: ALEApplication doesn't exist anymore. It has been renamed to MeshMovingApplication.

File: .gitignore
Comment (loumalouomega) [inline .gitignore]: You can not modify gitignore, please use the .git/info/exclude instead-> is personal

**Review by loumalouomega — CHANGES_REQUESTED** [review]
This is not in conditions for merging, right @roigcarlo ?

File: applications/FluidDynamicsApplication/python_scripts/navier_stokes_solver_vmsmonolithic.py
Comment (philbucher) [inline applications/FluidDynamicsApplication/python_scripts/navier_stokes_solver_vmsmonolithic.py]: You cannot do this, this breaks the fluid!

File: applications/ExternalSolversApplication/custom_external_libraries/SuperLU_4.3/FORTRAN/f77exm.out
Comment (loumalouomega) [inline applications/ExternalSolversApplication/custom_external_libraries/SuperLU_4.3/FORTRAN/f77exm.out]: Why do you remove this?

File: applications/MeshingApplication/custom_python/add_meshers_to_python.cpp
Comment (philbucher) [inline applications/MeshingApplication/custom_python/add_meshers_to_python.cpp]: please see the [Style Guide](https://github.com/KratosMultiphysics/Kratos/wiki/Style-Guide) for how to choose names => should be `ReGenerateMeshDroplet`

File: kratos/includes/variables.h
Comment (rubenzorrilla) [inline kratos/includes/variables.h]: Just asking, If this variables are related to the ULF application, why are you adding them in the core variables?

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]

File: kratos/python/add_containers_to_python.cpp
Comment (philbucher) [inline kratos/python/add_containers_to_python.cpp]: Why do you add these variables to the Core?

They are only used in your application from what I can see

File: applications/applications_interface.py
Comment (philbucher) [inline applications/applications_interface.py:143]: Naming conventions for applications is UpperCamleCase

Btw I would recommend to make a separate Pull request for adding a new application, this title of this PR is misleading

**Review by philbucher — CHANGES_REQUESTED** [review]
See the comments

File: .gitignore
Comment (philbucher) [reply .gitignore]: We should remove this ...

File: kratos/includes/variables.h
Comment (elafnm) [reply kratos/includes/variables.h]: Greetings dear @rubenzorrilla,

I had these variables to be used in the meshing application, and I found some issues during compiling kratos when I defined them locally. Accordingly and it was advised by Mr. @roigcarlo that:
"If you need that variable to be used in the meshing we should move the definition to the core"

regards,

Comment (philbucher) [reply kratos/includes/variables.h]: @elafnm this was like this until two months ago

Now you can define the same variable in several applications
i.e. you can now define the variables you need in the MeshingApp also there and then use them there

Comment (roigcarlo) [reply kratos/includes/variables.h]: Yup, now you can leave them in the application

File: applications/FluidDynamicsApplication/python_scripts/navier_stokes_solver_vmsmonolithic.py
Comment (playpaolo) [reply applications/FluidDynamicsApplication/python_scripts/navier_stokes_solver_vmsmonolithic.py]: Who did these changes? Is it me or @elafnm ?

File: .gitignore
Comment (playpaolo) [reply .gitignore]: None of these work I think. I mean these test examples.

File: applications/applications_interface.py
Comment (playpaolo) [reply applications/applications_interface.py:143]: @philbucher so do you suggest to put these files in the stash and do the revert?

Comment (philbucher) [reply applications/applications_interface.py:143]: I guess I would create a branch from this one and revert all unwanted changes such that the new Branch contains only the `FluidRveLagrangeMultipliersApplication`

Not the expert in git though, there might be better options

File: applications/FluidDynamicsApplication/python_scripts/navier_stokes_solver_vmsmonolithic.py
Comment (elafnm) [reply applications/FluidDynamicsApplication/python_scripts/navier_stokes_solver_vmsmonolithic.py]: Greetings dear @playpaolo and @rubenzorrilla 

I am not sure honestly, maybe I modified it by mistake. I will revert this change when I will be adding/pushing all my changes.

best,

File: .gitignore
Comment (elafnm) [reply .gitignore]: done dear,

File: applications/ExternalSolversApplication/custom_external_libraries/SuperLU_4.3/FORTRAN/f77exm.out
Comment (elafnm) [reply applications/ExternalSolversApplication/custom_external_libraries/SuperLU_4.3/FORTRAN/f77exm.out]: greetings dear,

the file was added as per the master,

File: .gitignore
Comment (elafnm) [reply .gitignore]: greetings dear,

done,

best regards,

File: applications/MeshingApplication/test_examples/dam3d_remesh.gid/out.out
Comment (elafnm) [reply applications/MeshingApplication/test_examples/dam3d_remesh.gid/out.out:2]: greetings dear,

the file was modified as per the master,

File: applications/applications_interface.py
Comment (elafnm) [reply applications/applications_interface.py:143]: greetings dear,

 the fluid_rve application was removed from the branch by Dr. Pavel,

best regards,

File: applications/ExternalSolversApplication/custom_external_libraries/SuperLU_4.3/FORTRAN/f77exm.out
Comment (elafnm) [reply applications/ExternalSolversApplication/custom_external_libraries/SuperLU_4.3/FORTRAN/f77exm.out]: Greetings dears @loumalouomega, @rubenzorrilla, and @philbucher,

kindly please, we would be very pleased and would highly appreciate your usual support by looking at the recommended changes,

Thanks and best regards,

Comment (loumalouomega) [reply applications/ExternalSolversApplication/custom_external_libraries/SuperLU_4.3/FORTRAN/f77exm.out]: I still don't understand why do you change this file

File: applications/ULFapplication/custom_processes/calculate_curvature.h
Comment (loumalouomega) [inline applications/ULFapplication/custom_processes/calculate_curvature.h:275]: I recommend to use flags, is faster and you don't need an extra variable

File: applications/ULFapplication/custom_processes/find_nodal_h_process.h
Comment (loumalouomega) [inline applications/ULFapplication/custom_processes/find_nodal_h_process.h]: Please, use the one in the core. We are strongly against code duplication

File: applications/ULFapplication/custom_processes/find_triple_point.h
Comment (loumalouomega) [inline applications/ULFapplication/custom_processes/find_triple_point.h:132]: Please use flags instead

File: applications/ExternalSolversApplication/custom_external_libraries/SuperLU_4.3/FORTRAN/f77exm.out
Comment (elafnm) [reply applications/ExternalSolversApplication/custom_external_libraries/SuperLU_4.3/FORTRAN/f77exm.out]: Greetings der @loumalouomega,

honestly dear, I am not sure if I changed it, 

dear @playpaolo, do you have an idea please about this file and if it was changed by you!

thanks and best regards,

File: applications/ULFapplication/custom_python/add_custom_utilities_to_python.cpp
Comment (loumalouomega) [inline applications/ULFapplication/custom_python/add_custom_utilities_to_python.cpp:59]: I am quite sure many of these includes are not necessary

File: applications/ULFapplication/python_scripts/SurfaceTension_monolithic_solver.py
Comment (loumalouomega) [inline applications/ULFapplication/python_scripts/SurfaceTension_monolithic_solver.py:4]: This solver is nor following the others applications "standard"

File: applications/ULFapplication/python_scripts/SurfaceTension_monolithic_solver.py
Comment (loumalouomega) [inline applications/ULFapplication/python_scripts/SurfaceTension_monolithic_solver.py:358]: Use the VariableUtils(), is faster and more efficient

File: applications/ULFapplication/python_scripts/runge_kutta_frac_step_solver.py
Comment (loumalouomega) [inline applications/ULFapplication/python_scripts/runge_kutta_frac_step_solver.py]: This is not following the standard

File: applications/ULFapplication/python_scripts/runge_kutta_frac_step_solver.py
Comment (loumalouomega) [inline applications/ULFapplication/python_scripts/runge_kutta_frac_step_solver.py:28]: Use VariableUtils()

File: applications/ULFapplication/python_scripts/runge_kutta_frac_step_solver.py
Comment (loumalouomega) [inline applications/ULFapplication/python_scripts/runge_kutta_frac_step_solver.py]: This is super inefficient, please use VariableUtils()

File: applications/ULFapplication/python_scripts/runge_kutta_frac_step_solver.py
Comment (loumalouomega) [inline applications/ULFapplication/python_scripts/runge_kutta_frac_step_solver.py]: many hardcoded values, how do I know what it is the meaning of these?, please use Parameters

File: applications/ULFapplication/python_scripts/ulf_strategy_PGLASS.py
Comment (loumalouomega) [inline applications/ULFapplication/python_scripts/ulf_strategy_PGLASS.py:38]: Strategies should be in cpp

File: applications/ULFapplication/test_examples/bottle_blowing/ProjectParameters.py
Comment (loumalouomega) [inline applications/ULFapplication/test_examples/bottle_blowing/ProjectParameters.py:4]: ProjectParameters should be in *.json format

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Much better, but still some work required

File: applications/ULFapplication/test_examples/bottle_blowing/ProjectParameters.py
Comment (elafnm) [reply applications/ULFapplication/test_examples/bottle_blowing/ProjectParameters.py:4]: Greetings dear @loumalouomega,

Thank you very much for your valuable and appreciated feedback, 

I will be discussing it with Dr. Pavel and do the needed changes accordingly,

Thanks and best regards,

File: applications/ExternalSolversApplication/custom_external_libraries/SuperLU_4.3/FORTRAN/f77exm.out
Comment (loumalouomega) [reply applications/ExternalSolversApplication/custom_external_libraries/SuperLU_4.3/FORTRAN/f77exm.out]: This looks like a minor change, just curious, check my comments for the rest of the review

File: applications/ULFapplication/test_examples/bottle_blowing/ProjectParameters.py
Comment (loumalouomega) [reply applications/ULFapplication/test_examples/bottle_blowing/ProjectParameters.py:4]: Best regards

Comment (playpaolo) [reply applications/ULFapplication/test_examples/bottle_blowing/ProjectParameters.py:4]: Hi @loumalouomega ! The problemtype used in ULF application is not preparing json files. We can think of updating this, but not within this pull request. Note that in this pull request new functionality is added. We will work on the format in the future.

File: applications/ULFapplication/python_scripts/runge_kutta_frac_step_solver.py
Comment (playpaolo) [reply applications/ULFapplication/python_scripts/runge_kutta_frac_step_solver.py]: Thanks. We will change it. @juliobarna  could you please take care of this one?

File: applications/ULFapplication/custom_processes/calculate_curvature.h
Comment (elafnm) [reply applications/ULFapplication/custom_processes/calculate_curvature.h:275]: Greetings dear @loumalouomega and very good day to you,

I would like to get your kind recommendation regarding using/defining flags locally,

As far as I am concerned, in our first pull request, when we were adding this file, https://github.com/KratosMultiphysics/Kratos/pull/1410, a similar flag issue was raised, and at that time it was not convenient to define our new flags in the core and I received the following feedback accordingly:

"If you don't think they are convenient for you, then just keep the current implementation. I also agree that adding a new local flag is not a good solution in this case"

so my concern here please, in the case now the flag variables can be defined locally, would you kindly please recommend me to follow a specific example/standard where these flags were defined and used locally.

thanks and best regards,

Comment (loumalouomega) [reply applications/ULFapplication/custom_processes/calculate_curvature.h:275]: The list of flags you can use is here:

https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/kratos_flags.h

If you need a new one you need to propose in an independent PR. The flags are limited to 64 and we can not add them happily. The flags you can use, according to you code could be:

FLUID
BOUNDARY
INTERFACE
FREE_SURFACE

Comment (loumalouomega) [reply applications/ULFapplication/custom_processes/calculate_curvature.h:275]: The only one you need to be careful is with ACTIVE. The nodes and conditions set as non ACTIVE will not be computed in the core builder and solvers

File: applications/ULFapplication/custom_processes/find_triple_point.h
Comment (elafnm) [reply applications/ULFapplication/custom_processes/find_triple_point.h:132]: Greeting dear @loumalouomega and very good day to you,

I just discussed this item with Dr. Alex, we can use 

KRATOS_CREATE_FLAG(STRUCTURE, 63);
 
but we need two more variables one for the triple points, and one for the Lagrangian-inlet

"If you need a new one you need to propose in an independent PR. "
so shall we go ahead with this pull request as is so that the code will be compiled, and then we start working on it in an independent PR!

thanks and best regard,

Comment (loumalouomega) [reply applications/ULFapplication/custom_processes/find_triple_point.h:132]: There is already one called INLET, the number of flags is limited and maybe we can not have all the ones one would like

Comment (elafnm) [reply applications/ULFapplication/custom_processes/find_triple_point.h:132]: thanks dear, so we will be needing one for the "triple_point".

best regards,

Comment (elafnm) [reply applications/ULFapplication/custom_processes/find_triple_point.h:132]: shall we go ahead with this pull request and then create a new pull request, 

or shall we include it in this pull request.

best regards,

Comment (loumalouomega) [reply applications/ULFapplication/custom_processes/find_triple_point.h:132]: Yes, change the ones that you can currently change with the existing flags

File: applications/ULFapplication/python_scripts/ulf_strategy_PGLASS.py
Comment (playpaolo) [reply applications/ULFapplication/python_scripts/ulf_strategy_PGLASS.py:38]: @loumalouomega I think we are having some sort of conceptual misunderstanding here. We are not planning to change the structure of the entire ULFApplication in this pull request (when the application was originally added we were not asked to reproduce the format of say FluidDynamicsApplication or other "user-oriented" standardized application). All the python solvers in ULF are divided into two parts: solver and strategy. We are neither interested in switching to JSON at the moment as we are using a different problem type.

**Comment by loumalouomega** [comment]
OK, then I will approve, but wait for the approval of the others

**Review by loumalouomega — DISMISSED** [review]
Let for the approval of the others

File: applications/ULFapplication/custom_processes/find_triple_point.h
Comment (playpaolo) [reply applications/ULFapplication/custom_processes/find_triple_point.h:132]: @loumalouomega I am strongly against doing this in this pull request. Flags you are talking about involve changes in 75% of our application and also the meshers we use from MeshingApplication. I suggest to make a separate pull request for "moving" all our old flag variables (IS_FREE_SURFACE, IS_STRUCTURE, IS_TRIPPLE_POINT) to new format. But lets do it in a separate pull request, otherwise things will get mixed.

Comment (loumalouomega) [reply applications/ULFapplication/custom_processes/find_triple_point.h:132]: I understand, I have approved

Comment (philbucher) [reply applications/ULFapplication/custom_processes/find_triple_point.h:132]: I agree with not using flags in this PR

nevertheless I recommend to NOT compare doubles with `==` but rather sth like
`if abs(val1-val2) < 1e-15` (just an example)

Quoting https://floating-point-gui.de/errors/comparison/:
>Due to rounding errors, most floating-point numbers end up being slightly imprecise. As long as this imprecision stays small, it can usually be ignored. However, it also means that numbers expected to be equal (e.g. when calculating the same result through different correct methods) often differ slightly, and a simple equality test fails

I will remove my blocker since my concerns were addressed

**Review by philbucher — DISMISSED** [review]
Actually the changes I requested initially are again there, even though I think they were solved at some point 
e.g. changes to gitignore, the change in the fluid solver and adding the variables in the core instead of the application

@elafnm @playpaolo can you check please?

**Comment by playpaolo** [comment]
Hi @philbucher and @loumalouomega ! One question..  How can we bring in "standard" .gitignore to our branch in case by mistake we pushed a modified .gitignore at some point?

**Comment by philbucher** [comment]
I think in this case it would be the easiest to copy paste the one from master and commit it, then the changes should disappear

**Comment by loumalouomega** [comment]
Indeed @philbucher

**Comment by playpaolo** [comment]
@philbucher and @loumalouomega  could you please check if it is fixed now? thank you!

File: applications/MeshingApplication/custom_python/add_meshers_to_python.cpp
Comment (philbucher) [reply applications/MeshingApplication/custom_python/add_meshers_to_python.cpp]: @playpaolo this is still remaining

File: .gitignore
Comment (philbucher) [inline .gitignore]: @playpaolo please don't remove this

File: kratos/python/add_containers_to_python.cpp
Comment (philbucher) [reply kratos/python/add_containers_to_python.cpp]: @playpaolo this was also changed at some point

**Review by philbucher — COMMENTED** [review]
@playpaolo the three things I pointed out are still missing

**Comment by playpaolo** [comment]
@elafnm could you please see the thing about the variables @philbucher  is talking about? please put all your variables in the application and not in the core of kratos.

**Comment by loumalouomega** [comment]
I reverted last problematic files, approving

**Review by loumalouomega — DISMISSED** [review]
Done

**Comment by playpaolo** [comment]
Thank you @loumalouomega !

**Comment by philbucher** [comment]
now from my side only the moving of the variables from the core to the application is left

**Comment by elafnm** [comment]
greetings dears @playpaolo and @philbucher and very good day to you,

done, they were moved from the core and defined locally. Please let me know if you have any further concern and it will be my pleasure to correct it accordingly.

Thanks and best regards,

**Review by loumalouomega — APPROVED** [review]

**Comment by elafnm** [comment]
greetings dears @philbucher and @rubenzorrilla and very good day to you,

could you please have a look at the changes done as per your recommendations; and please let me know in case you have any further concerns.

best regards,

**Comment by philbucher** [comment]
@elafnm I am fine with your changes, I removed my blocker

It is up to @rubenzorrilla now

**Comment by elafnm** [comment]
thank you very much dear @philbucher,

best regards,

File: kratos/includes/variables.h
Comment (rubenzorrilla) [inline kratos/includes/variables.h:415]: In the future avoid these kind of aesthetic changes in such a big PR. (probably is something related with the configuration of your code editor). These things make the review process much tedious.

File: kratos/sources/variables.cpp
Comment (rubenzorrilla) [inline kratos/sources/variables.cpp:962]: Same as in `variables.h`

**Review by rubenzorrilla — APPROVED** [review]
Thanks for taking into account our demands. Since all the changes are local in your application I'm approving.

As a comment aside, it's better to do more PRs with less changes. In this way the review will be much faster and you won't need to wait for so long for merging with the master.

**Comment by elafnm** [comment]
Thank you very much dears @playpaolo, @rubenzorrilla, @philbucher, @roigcarlo, and @loumalouomega for your supports and approvals,

sure will follow your recommendations for future PRs,

with our best regards,

**Pull request merged**
